cmd/maya-apiserver/spc-watcher/lease.go

```diff
@@ -86,18 +177,79 @@ func (sl *spcLease) UpdateLease(leaseValue string) (*apis.StoragePoolClaim, erro
 	return spcObject, nil
 }
 
-// Can be used for reconcile loop use cases
-// TODO Remove using patch instead of update
-func (sl *spcLease) RemoveLease() *apis.StoragePoolClaim {
-	spcObject, err := sl.UpdateLease("")
+func (sl *spcLease) RemoveLease() {
+	_, err := sl.patchSpc()
 	if err != nil {
-		runtime.HandleError(fmt.Errorf("Lease could not be removed:%v", err))
+		newErr := fmt.Errorf("Lease could not be removed:%v", err)
+		runtime.HandleError(newErr)
 	}
-	return spcObject
+	glog.Info("Lease removed successfully on storagepoolclaim")
 }
 
 func (sl *spcLease) getPodName() string {
-	podName := menv.Get(menv.OpenEBSMayaPodName)
-	podNameSpace := menv.Get(menv.OpenEBSNamespace)
+	podName := env.Get(env.OpenEBSMayaPodName)
+	podNameSpace := env.Get(env.OpenEBSNamespace)
 	return podNameSpace + "/" + podName
 }
+
+// patchSpc will patch the spc object to release the lease
+func (sl *spcLease) patchSpc() (*apis.StoragePoolClaim, error) {
+	spcPatch := make([]SPCPatch, 1)
+	// setting operation as remove
+	spcPatch[0].Op = PatchOperation
+	// object to be removed is finalizers
+	spcPatch[0].Path = PatchPath
+	leaseValueObj, err := parseLeaseValue(sl.spcObject.Annotations[SpcLeaseKey])
+	leaseValueObj.HolderIdentity = ""
+	newLeaseValue, err := json.Marshal(leaseValueObj)
+	if err != nil {
+		return nil, err
+	}
+	spcPatch[0].Value = string(newLeaseValue)
+	spcPatchJSON, err := json.Marshal(spcPatch)
+	if err != nil {
+		glog.Errorf("Error marshalling spcPatch object: %s", err)
+	}
+	obj, err := sl.oecs.OpenebsV1alpha1().StoragePoolClaims().Patch(sl.spcObject.Name, types.JSONPatchType, spcPatchJSON)
+	return obj, err
+}
+
+// checkLeaderLiveness checks whether the holder of lease is live or not
+// If the holder is not live or does not exists the function will return true.
+
+// If the holder of lease is not live or does not exists the lease can be acquired
+// by the other contestant(i.e. maya pod)
+func (sl *spcLease) checkLeaderLiveness(leaseValueObj lease) bool {
+	holderName := leaseValueObj.HolderIdentity
+	podDetails := strings.Split(holderName, "/")
+	// Check whether the holder is live or not
+	pod, _ := sl.kubeclientset.CoreV1().Pods(podDetails[0]).Get(podDetails[1], meta_v1.GetOptions{})
+	if pod == nil {
+		return true
+	}
+	podStatus := pod.Status.Phase
+	if string(podStatus) != string(v1.PodRunning) {
+		return true
+	}
+	return false
+}
+
+// A lease is expired if it has a empty holder name.
+// The holder of lease can only expire the lease.
+// isLeaseExpired return true if the lease is expired.
+func isLeaseExpired(leaseValueObj lease) bool {
```

Comment:
This should be a method of Lease interface

Reply:
I am keeping this out of Interface as Interface will have only Hold, Update and Release functions that can be implemented for different kind of object.
Now different kind of object can have different logic of lease acquisition so IMO we should not mandate the way of implementation. There might be cases where we do not want to use isLeaseExpired() function for certain type of lease use cases.
Additionally, the lease algo here is typically tight bound to SPC and it is necessary as the functions e.g. Hold(), Update() and Release() need to get the object of the type not just the Interface type.